// @dvirtz/parquets/lib/codec/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PARQUET_CODEC = void 0;
const RLE = require("./rle");
const PLAIN = require("./plain");
const PLAIN_DICTIONARY = require("./plain_dictionary");
__exportStar(require("./declare"), exports);
exports.PARQUET_CODEC = {
    PLAIN: {
        encodeValues: PLAIN.encodeValues,
        decodeValues: PLAIN.decodeValues
    },
    RLE: {
        encodeValues: RLE.encodeValues,
        decodeValues: RLE.decodeValues
    },
    PLAIN_DICTIONARY: {
        encodeValues: PLAIN_DICTIONARY.encodeValues,
        decodeValues: PLAIN_DICTIONARY.decodeValues
    }
};
