// @dvirtz/parquets/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParquetShredder = void 0;
const ParquetShredder = require("./shred");
exports.ParquetShredder = ParquetShredder;
__exportStar(require("./declare"), exports);
var reader_1 = require("./reader");
Object.defineProperty(exports, "ParquetCursor", { enumerable: true, get: function () { return reader_1.ParquetCursor; } });
Object.defineProperty(exports, "ParquetEnvelopeReader", { enumerable: true, get: function () { return reader_1.ParquetEnvelopeReader; } });
Object.defineProperty(exports, "ParquetReader", { enumerable: true, get: function () { return reader_1.ParquetReader; } });
var schema_1 = require("./schema");
Object.defineProperty(exports, "ParquetSchema", { enumerable: true, get: function () { return schema_1.ParquetSchema; } });
var writer_1 = require("./writer");
Object.defineProperty(exports, "ParquetEnvelopeWriter", { enumerable: true, get: function () { return writer_1.ParquetEnvelopeWriter; } });
Object.defineProperty(exports, "ParquetTransformer", { enumerable: true, get: function () { return writer_1.ParquetTransformer; } });
Object.defineProperty(exports, "ParquetWriter", { enumerable: true, get: function () { return writer_1.ParquetWriter; } });
