// @finos/perspective/dist/esm/psp.async.js
import "core-js/modules/es6.regexp.constructor";
import "core-js/modules/es6.regexp.split";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.regexp.flags";
import "core-js/modules/es6.regexp.replace";
import "core-js/modules/es6.regexp.to-string";

var load_perspective = function () {
  var _scriptDir = typeof document !== 'undefined' && document.currentScript ? document.currentScript.src : undefined;

  return function (load_perspective) {
    load_perspective = load_perspective || {};
    var e;
    e || (e = typeof load_perspective !== 'undefined' ? load_perspective : {});
    var aa = {},
        k;

    for (k in e) e.hasOwnProperty(k) && (aa[k] = e[k]);

    var ba = [],
        ca = "./this.program";

    function da(a, b) {
      throw b;
    }

    var q = !1,
        r = !1,
        ea = !1,
        ha = !1,
        ia = !1;
    q = "object" === typeof window;
    r = "function" === typeof importScripts;
    ea = (ha = "object" === typeof process && "object" === typeof process.versions && "string" === typeof process.versions.node) && !q && !r;
    ia = !q && !ea && !r;
    var t = "",
        ja,
        ka;

    if (ea) {
      t = __dirname + "/";
      var la, ma;

      ja = function ja(a, b) {
        la || (la = require("fs"));
        ma || (ma = require("path"));
        a = ma.normalize(a);
        a = la.readFileSync(a);
        return b ? a : a.toString();
      };

      ka = function ka(a) {
        a = ja(a, !0);
        a.buffer || (a = new Uint8Array(a));
        a.buffer || u("Assertion failed: undefined");
        return a;
      };

      1 < process.argv.length && (ca = process.argv[1].replace(/\\/g, "/"));
      ba = process.argv.slice(2);
      process.on("uncaughtException", function (a) {
        if (!(a instanceof na)) throw a;
      });
      process.on("unhandledRejection", u);

      da = function da(a) {
        process.exit(a);
      };

      e.inspect = function () {
        return "[Emscripten Module object]";
      };
    } else if (ia) "undefined" != typeof read && (ja = function ja(a) {
      return read(a);
    }), ka = function ka(a) {
      if ("function" === typeof readbuffer) return new Uint8Array(readbuffer(a));
      a = read(a, "binary");
      "object" === typeof a || u("Assertion failed: undefined");
      return a;
    }, "undefined" != typeof scriptArgs ? ba = scriptArgs : "undefined" != typeof arguments && (ba = arguments), "function" === typeof quit && (da = function da(a) {
      quit(a);
    }), "undefined" !== typeof print && ("undefined" === typeof console && (console = {}), console.log = print, console.warn = console.error = "undefined" !== typeof printErr ? printErr : print);else if (q || r) r ? t = self.location.href : document.currentScript && (t = document.currentScript.src), _scriptDir && (t = _scriptDir), 0 !== t.indexOf("blob:") ? t = t.substr(0, t.lastIndexOf("/") + 1) : t = "", ja = function ja(a) {
      var b = new XMLHttpRequest();
      b.open("GET", a, !1);
      b.send(null);
      return b.responseText;
    }, r && (ka = function ka(a) {
      var b = new XMLHttpRequest();
      b.open("GET", a, !1);
      b.responseType = "arraybuffer";
      b.send(null);
      return new Uint8Array(b.response);
    });

    var oa = e.print || console.log.bind(console),
        v = e.printErr || console.warn.bind(console);

    for (k in aa) aa.hasOwnProperty(k) && (e[k] = aa[k]);

    aa = null;
    e.arguments && (ba = e.arguments);
    e.thisProgram && (ca = e.thisProgram);
    e.quit && (da = e.quit);

    function pa(a) {
      var b = x[qa >> 2];
      a = b + a + 15 & -16;
      a > ra() && u();
      x[qa >> 2] = a;
      return b;
    }

    var sa;
    e.wasmBinary && (sa = e.wasmBinary);
    var ta;
    e.noExitRuntime && (ta = e.noExitRuntime);
    "object" !== typeof WebAssembly && v("no native wasm support detected");
    var y,
        ua = new WebAssembly.Table({
      initial: 2034,
      maximum: 2034,
      element: "anyfunc"
    }),
        va = !1,
        wa = "undefined" !== typeof TextDecoder ? new TextDecoder("utf8") : void 0;

    function xa(a, b, c) {
      var d = b + c;

      for (c = b; a[c] && !(c >= d);) ++c;

      if (16 < c - b && a.subarray && wa) return wa.decode(a.subarray(b, c));

      for (d = ""; b < c;) {
        var f = a[b++];

        if (f & 128) {
          var g = a[b++] & 63;
          if (192 == (f & 224)) d += String.fromCharCode((f & 31) << 6 | g);else {
            var h = a[b++] & 63;
            f = 224 == (f & 240) ? (f & 15) << 12 | g << 6 | h : (f & 7) << 18 | g << 12 | h << 6 | a[b++] & 63;
            65536 > f ? d += String.fromCharCode(f) : (f -= 65536, d += String.fromCharCode(55296 | f >> 10, 56320 | f & 1023));
          }
        } else d += String.fromCharCode(f);
      }

      return d;
    }

    function z(a) {
      return a ? xa(A, a, void 0) : "";
    }

    function ya(a, b, c, d) {
      if (0 < d) {
        d = c + d - 1;

        for (var f = 0; f < a.length; ++f) {
          var g = a.charCodeAt(f);

          if (55296 <= g && 57343 >= g) {
            var h = a.charCodeAt(++f);
            g = 65536 + ((g & 1023) << 10) | h & 1023;
          }

          if (127 >= g) {
            if (c >= d) break;
            b[c++] = g;
          } else {
            if (2047 >= g) {
              if (c + 1 >= d) break;
              b[c++] = 192 | g >> 6;
            } else {
              if (65535 >= g) {
                if (c + 2 >= d) break;
                b[c++] = 224 | g >> 12;
              } else {
                if (c + 3 >= d) break;
                b[c++] = 240 | g >> 18;
                b[c++] = 128 | g >> 12 & 63;
              }

              b[c++] = 128 | g >> 6 & 63;
            }

            b[c++] = 128 | g & 63;
          }
        }

        b[c] = 0;
      }
    }

    function za(a) {
      for (var b = 0, c = 0; c < a.length; ++c) {
        var d = a.charCodeAt(c);
        55296 <= d && 57343 >= d && (d = 65536 + ((d & 1023) << 10) | a.charCodeAt(++c) & 1023);
        127 >= d ? ++b : b = 2047 >= d ? b + 2 : 65535 >= d ? b + 3 : b + 4;
      }

      return b;
    }

    "undefined" !== typeof TextDecoder && new TextDecoder("utf-16le");

    function Aa(a) {
      var b = za(a) + 1,
          c = Ba(b);
      ya(a, B, c, b);
      return c;
    }

    function Ca(a) {
      0 < a % 65536 && (a += 65536 - a % 65536);
      return a;
    }

    var buffer, B, A, Da, Ea, x, C, Fa, Ga;

    function Ha(a) {
      buffer = a;
      e.HEAP8 = B = new Int8Array(a);
      e.HEAP16 = Da = new Int16Array(a);
      e.HEAP32 = x = new Int32Array(a);
      e.HEAPU8 = A = new Uint8Array(a);
      e.HEAPU16 = Ea = new Uint16Array(a);
      e.HEAPU32 = C = new Uint32Array(a);
      e.HEAPF32 = Fa = new Float32Array(a);
      e.HEAPF64 = Ga = new Float64Array(a);
    }

    var qa = 95376,
        Ia = e.TOTAL_MEMORY || 16777216;
    e.wasmMemory ? y = e.wasmMemory : y = new WebAssembly.Memory({
      initial: Ia / 65536
    });
    y && (buffer = y.buffer);
    Ia = buffer.byteLength;
    Ha(buffer);
    x[qa >> 2] = 5338448;

    function Ja(a) {
      for (; 0 < a.length;) {
        var b = a.shift();
        if ("function" == typeof b) b();else {
          var c = b.U;
          "number" === typeof c ? void 0 === b.G ? e.dynCall_v(c) : e.dynCall_vi(c, b.G) : c(void 0 === b.G ? null : b.G);
        }
      }
    }

    var Ka = [],
        La = [],
        Ma = [],
        Na = [],
        Oa = [],
        Pa = !1;

    function Qa() {
      var a = e.preRun.shift();
      Ka.unshift(a);
    }

    var D = 0,
        Ra = null,
        Sa = null;
    e.preloadedImages = {};
    e.preloadedAudios = {};

    function u(a) {
      if (e.onAbort) e.onAbort(a);
      oa(a);
      v(a);
      va = !0;
      throw "abort(" + a + "). Build with -s ASSERTIONS=1 for more info.";
    }

    function Ta() {
      var a = E;
      return String.prototype.startsWith ? a.startsWith("data:application/octet-stream;base64,") : 0 === a.indexOf("data:application/octet-stream;base64,");
    }

    var E = "psp.async.wasm";

    if (!Ta()) {
      var Ua = E;
      E = e.locateFile ? e.locateFile(Ua, t) : t + Ua;
    }

    function Va() {
      try {
        if (sa) return new Uint8Array(sa);
        if (ka) return ka(E);
        throw "both async and sync fetching of the wasm failed";
      } catch (a) {
        u(a);
      }
    }

    function Wa() {
      return sa || !q && !r || "function" !== typeof fetch ? new Promise(function (a) {
        a(Va());
      }) : fetch(E, {
        credentials: "same-origin"
      }).then(function (a) {
        if (!a.ok) throw "failed to load wasm binary file at '" + E + "'";
        return a.arrayBuffer();
      }).catch(function () {
        return Va();
      });
    }

    var Xa = [function () {
      throw Error("abort()");
    }, function () {
      if ("undefined" !== typeof self) try {
        if (self.dispatchEvent && !self.m && null !== self.document) {
          self.m = !0;
          var a = self.document.createEvent("Event");
          a.initEvent("perspective-ready", !1, !0);
          self.dispatchEvent(a);
        } else !self.document && self.postMessage && self.postMessage({});
      } catch (b) {}
    }];
    La.push({
      U: function U() {
        Ya();
      }
    });

    function Za(a) {
      return a.replace(/\b_Z[\w\d_]+/g, function (a) {
        return a === a ? a : a + " [" + a + "]";
      });
    }

    function $a() {
      var a = Error();

      if (!a.stack) {
        try {
          throw Error(0);
        } catch (b) {
          a = b;
        }

        if (!a.stack) return "(no stack trace available)";
      }

      return a.stack.toString();
    }

    var F = {};

    function ab(a) {
      if (ab.m) {
        var b = x[a >> 2];
        var c = x[b >> 2];
      } else ab.m = !0, F.USER = F.LOGNAME = "web_user", F.PATH = "/", F.PWD = "/", F.HOME = "/home/web_user", F.LANG = ("object" === typeof navigator && navigator.languages && navigator.languages[0] || "C").replace("-", "_") + ".UTF-8", F._ = ca, c = Pa ? G(1024) : pa(1024), b = Pa ? G(256) : pa(256), x[b >> 2] = c, x[a >> 2] = b;

      a = [];
      var d = 0,
          f;

      for (f in F) if ("string" === typeof F[f]) {
        var g = f + "=" + F[f];
        a.push(g);
        d += g.length;
      }

      if (1024 < d) throw Error("Environment size exceeded TOTAL_ENV_SIZE!");

      for (f = 0; f < a.length; f++) {
        d = g = a[f];

        for (var h = c, l = 0; l < d.length; ++l) B[h++ >> 0] = d.charCodeAt(l);

        B[h >> 0] = 0;
        x[b + 4 * f >> 2] = c;
        c += g.length + 1;
      }

      x[b + 4 * a.length >> 2] = 0;
    }

    function bb(a, b) {
      Na.unshift({
        U: a,
        G: b
      });
    }

    function cb(a) {
      e.___errno_location && (x[e.___errno_location() >> 2] = a);
      return a;
    }

    var db = [null, [], []];

    function eb(a, b) {
      var c = db[a];
      0 === b || 10 === b ? ((1 === a ? oa : v)(xa(c, 0)), c.length = 0) : c.push(b);
    }

    var H = 0;

    function I() {
      H += 4;
      return x[H - 4 >> 2];
    }

    var J = {};

    function fb(a, b, c) {
      a |= 0;
      c |= 0;
      var d;
      var f = a + c | 0;
      b = (b | 0) & 255;

      if (67 <= (c | 0)) {
        for (; 0 != (a & 3);) B[a >> 0] = b, a = a + 1 | 0;

        var g = f & -4 | 0;
        var h = b | b << 8 | b << 16 | b << 24;

        for (d = g - 64 | 0; (a | 0) <= (d | 0);) x[a >> 2] = h, x[a + 4 >> 2] = h, x[a + 8 >> 2] = h, x[a + 12 >> 2] = h, x[a + 16 >> 2] = h, x[a + 20 >> 2] = h, x[a + 24 >> 2] = h, x[a + 28 >> 2] = h, x[a + 32 >> 2] = h, x[a + 36 >> 2] = h, x[a + 40 >> 2] = h, x[a + 44 >> 2] = h, x[a + 48 >> 2] = h, x[a + 52 >> 2] = h, x[a + 56 >> 2] = h, x[a + 60 >> 2] = h, a = a + 64 | 0;

        for (; (a | 0) < (g | 0);) x[a >> 2] = h, a = a + 4 | 0;
      }

      for (; (a | 0) < (f | 0);) B[a >> 0] = b, a = a + 1 | 0;

      return f - c | 0;
    }

    function gb(a, b, c, d, f, g) {
      g <<= 12;
      if (0 !== (d & 16) && 0 !== a % 16384) return -22;

      if (0 !== (d & 32)) {
        a = hb(16384, b);
        if (!a) return -12;
        fb(a, 0, b);
        c = !0;
      } else {
        var h = (void 0).da(f);
        if (!h) return -9;
        c = (void 0).za(h, A, a, b, g, c, d);
        a = c.c;
        c = c.R;
      }

      J.W[a] = {
        ka: a,
        ja: b,
        R: c,
        fd: f,
        flags: d
      };
      return a;
    }

    function ib(a, b) {
      if (-1 === a || 0 === b) return -22;
      var c = J.W[a];
      if (!c) return 0;

      if (b === c.ja) {
        var d = (void 0).da(c.fd);
        J.wa(a, d, b, c.flags);
        (void 0).Aa(d);
        J.W[a] = null;
        c.R && K(c.ka);
      }

      return 0;
    }

    function jb(a, b, c, d) {
      try {
        for (var f = 0, g = 0; g < c; g++) {
          for (var h = x[b + 8 * g >> 2], l = x[b + (8 * g + 4) >> 2], n = 0; n < l; n++) eb(a, A[h + n]);

          f += l;
        }

        x[d >> 2] = f;
        return 0;
      } catch (m) {
        return u(m), -m.u;
      }
    }

    var kb = {};

    function lb(a) {
      for (; a.length;) {
        var b = a.pop();
        a.pop()(b);
      }
    }

    function L(a) {
      return this.fromWireType(C[a >> 2]);
    }

    var M = {},
        N = {},
        mb = {};

    function nb(a) {
      if (void 0 === a) return "_unknown";
      a = a.replace(/[^a-zA-Z0-9_]/g, "$");
      var b = a.charCodeAt(0);
      return 48 <= b && 57 >= b ? "_" + a : a;
    }

    function ob(a, b) {
      a = nb(a);
      return new Function("body", "return function " + a + '() {\n    "use strict";    return body.apply(this, arguments);\n};\n')(b);
    }

    function pb(a, b) {
      var c = ob(b, function (a) {
        this.name = b;
        this.message = a;
        a = Error(a).stack;
        void 0 !== a && (this.stack = this.toString() + "\n" + a.replace(/^Error(:[^\n]*)?\n/, ""));
      });
      c.prototype = Object.create(a.prototype);
      c.prototype.constructor = c;

      c.prototype.toString = function () {
        return void 0 === this.message ? this.name : this.name + ": " + this.message;
      };

      return c;
    }

    var qb = void 0;

    function rb(a) {
      throw new qb(a);
    }

    function O(a, b, c) {
      function d(b) {
        b = c(b);
        b.length !== a.length && rb("Mismatched type converter count");

        for (var d = 0; d < a.length; ++d) Q(a[d], b[d]);
      }

      a.forEach(function (a) {
        mb[a] = b;
      });
      var f = Array(b.length),
          g = [],
          h = 0;
      b.forEach(function (a, b) {
        N.hasOwnProperty(a) ? f[b] = N[a] : (g.push(a), M.hasOwnProperty(a) || (M[a] = []), M[a].push(function () {
          f[b] = N[a];
          ++h;
          h === g.length && d(f);
        }));
      });
      0 === g.length && d(f);
    }

    function sb(a) {
      switch (a) {
        case 1:
          return 0;

        case 2:
          return 1;

        case 4:
          return 2;

        case 8:
          return 3;

        default:
          throw new TypeError("Unknown type size: " + a);
      }
    }

    function tb() {
      for (var a = Array(256), b = 0; 256 > b; ++b) a[b] = String.fromCharCode(b);

      ub = a;
    }

    var ub = void 0;

    function R(a) {
      for (var b = ""; A[a];) b += ub[A[a++]];

      return b;
    }

    var vb = void 0;

    function S(a) {
      throw new vb(a);
    }

    function Q(a, b, c) {
      c = c || {};
      if (!("argPackAdvance" in b)) throw new TypeError("registerType registeredInstance requires argPackAdvance");
      var d = b.name;
      a || S('type "' + d + '" must have a positive integer typeid pointer');

      if (N.hasOwnProperty(a)) {
        if (c.ha) return;
        S("Cannot register type '" + d + "' twice");
      }

      N[a] = b;
      delete mb[a];
      M.hasOwnProperty(a) && (b = M[a], delete M[a], b.forEach(function (a) {
        a();
      }));
    }

    function wb(a) {
      if (!(this instanceof T && a instanceof T)) return !1;
      var b = this.a.f.b,
          c = this.a.c,
          d = a.a.f.b;

      for (a = a.a.c; b.o;) c = b.F(c), b = b.o;

      for (; d.o;) a = d.F(a), d = d.o;

      return b === d && c === a;
    }

    function xb(a) {
      return {
        count: a.count,
        w: a.w,
        C: a.C,
        c: a.c,
        f: a.f,
        i: a.i,
        l: a.l
      };
    }

    function yb(a) {
      S(a.a.f.b.name + " instance already deleted");
    }

    var zb = !1;

    function Ab() {}

    function Bb(a) {
      a.i ? a.l.s(a.i) : a.f.b.s(a.c);
    }

    function Cb(a) {
      --a.count.value;
      0 === a.count.value && Bb(a);
    }

    function Db(a) {
      if ("undefined" === typeof FinalizationGroup) return Db = function Db(a) {
        return a;
      }, a;
      zb = new FinalizationGroup(function (a) {
        for (var b = a.next(); !b.done; b = a.next()) b = b.value, b.c ? Cb(b) : console.warn("object already deleted: " + b.c);
      });

      Db = function Db(a) {
        zb.register(a, a.a, a.a);
        return a;
      };

      Ab = function Ab(a) {
        zb.unregister(a.a);
      };

      return Db(a);
    }

    function Eb() {
      this.a.c || yb(this);
      if (this.a.C) return this.a.count.value += 1, this;
      var a = Db(Object.create(Object.getPrototypeOf(this), {
        a: {
          value: xb(this.a)
        }
      }));
      a.a.count.value += 1;
      a.a.w = !1;
      return a;
    }

    function Fb() {
      this.a.c || yb(this);
      this.a.w && !this.a.C && S("Object already scheduled for deletion");
      Ab(this);
      Cb(this.a);
      this.a.C || (this.a.i = void 0, this.a.c = void 0);
    }

    function Gb() {
      return !this.a.c;
    }

    var Hb = void 0,
        Ib = [];

    function Jb() {
      for (; Ib.length;) {
        var a = Ib.pop();
        a.a.w = !1;
        a["delete"]();
      }
    }

    function Kb() {
      this.a.c || yb(this);
      this.a.w && !this.a.C && S("Object already scheduled for deletion");
      Ib.push(this);
      1 === Ib.length && Hb && Hb(Jb);
      this.a.w = !0;
      return this;
    }

    function Lb() {
      T.prototype.isAliasOf = wb;
      T.prototype.clone = Eb;
      T.prototype["delete"] = Fb;
      T.prototype.isDeleted = Gb;
      T.prototype.deleteLater = Kb;
    }

    function T() {}

    var Mb = {};

    function Nb(a, b, c) {
      if (void 0 === a[b].g) {
        var d = a[b];

        a[b] = function () {
          a[b].g.hasOwnProperty(arguments.length) || S("Function '" + c + "' called with an invalid number of arguments (" + arguments.length + ") - expects one of (" + a[b].g + ")!");
          return a[b].g[arguments.length].apply(this, arguments);
        };

        a[b].g = [];
        a[b].g[d.H] = d;
      }
    }

    function Ob(a, b, c) {
      e.hasOwnProperty(a) ? ((void 0 === c || void 0 !== e[a].g && void 0 !== e[a].g[c]) && S("Cannot register public name '" + a + "' twice"), Nb(e, a, a), e.hasOwnProperty(c) && S("Cannot register multiple overloads of a function with the same number of arguments (" + c + ")!"), e[a].g[c] = b) : (e[a] = b, void 0 !== c && (e[a].Ba = c));
    }

    function Pb(a, b, c, d, f, g, h, l) {
      this.name = a;
      this.constructor = b;
      this.A = c;
      this.s = d;
      this.o = f;
      this.aa = g;
      this.F = h;
      this.Z = l;
      this.ma = [];
    }

    function Qb(a, b, c) {
      for (; b !== c;) b.F || S("Expected null or instance of " + c.name + ", got an instance of " + b.name), a = b.F(a), b = b.o;

      return a;
    }

    function Rb(a, b) {
      if (null === b) return this.N && S("null is not a valid " + this.name), 0;
      b.a || S('Cannot pass "' + U(b) + '" as a ' + this.name);
      b.a.c || S("Cannot pass deleted object as a pointer of type " + this.name);
      return Qb(b.a.c, b.a.f.b, this.b);
    }

    function Sb(a, b) {
      if (null === b) {
        this.N && S("null is not a valid " + this.name);

        if (this.J) {
          var c = this.O();
          null !== a && a.push(this.s, c);
          return c;
        }

        return 0;
      }

      b.a || S('Cannot pass "' + U(b) + '" as a ' + this.name);
      b.a.c || S("Cannot pass deleted object as a pointer of type " + this.name);
      !this.I && b.a.f.I && S("Cannot convert argument of type " + (b.a.l ? b.a.l.name : b.a.f.name) + " to parameter type " + this.name);
      c = Qb(b.a.c, b.a.f.b, this.b);
      if (this.J) switch (void 0 === b.a.i && S("Passing raw pointer to smart pointer is illegal"), this.ra) {
        case 0:
          b.a.l === this ? c = b.a.i : S("Cannot convert argument of type " + (b.a.l ? b.a.l.name : b.a.f.name) + " to parameter type " + this.name);
          break;

        case 1:
          c = b.a.i;
          break;

        case 2:
          if (b.a.l === this) c = b.a.i;else {
            var d = b.clone();
            c = this.na(c, V(function () {
              d["delete"]();
            }));
            null !== a && a.push(this.s, c);
          }
          break;

        default:
          S("Unsupporting sharing policy");
      }
      return c;
    }

    function Tb(a, b) {
      if (null === b) return this.N && S("null is not a valid " + this.name), 0;
      b.a || S('Cannot pass "' + U(b) + '" as a ' + this.name);
      b.a.c || S("Cannot pass deleted object as a pointer of type " + this.name);
      b.a.f.I && S("Cannot convert argument of type " + b.a.f.name + " to parameter type " + this.name);
      return Qb(b.a.c, b.a.f.b, this.b);
    }

    function Ub(a) {
      this.X && (a = this.X(a));
      return a;
    }

    function Vb(a) {
      this.s && this.s(a);
    }

    function Wb(a) {
      if (null !== a) a["delete"]();
    }

    function Xb(a, b, c) {
      if (b === c) return a;
      if (void 0 === c.o) return null;
      a = Xb(a, b, c.o);
      return null === a ? null : c.Z(a);
    }

    function Yb() {
      return Object.keys(Zb).length;
    }

    function $b() {
      var a = [],
          b;

      for (b in Zb) Zb.hasOwnProperty(b) && a.push(Zb[b]);

      return a;
    }

    function ac(a) {
      Hb = a;
      Ib.length && Hb && Hb(Jb);
    }

    function bc() {
      e.getInheritedInstanceCount = Yb;
      e.getLiveInheritedInstances = $b;
      e.flushPendingDeletes = Jb;
      e.setDelayFunction = ac;
    }

    var Zb = {};

    function cc(a, b) {
      for (void 0 === b && S("ptr should not be undefined"); a.o;) b = a.F(b), a = a.o;

      return b;
    }

    function dc(a, b) {
      b = cc(a, b);
      return Zb[b];
    }

    function ec(a, b) {
      b.f && b.c || rb("makeClassHandle requires ptr and ptrType");
      !!b.l !== !!b.i && rb("Both smartPtrType and smartPtr must be specified");
      b.count = {
        value: 1
      };
      return Db(Object.create(a, {
        a: {
          value: b
        }
      }));
    }

    function fc(a) {
      function b() {
        return this.J ? ec(this.b.A, {
          f: this.la,
          c: c,
          l: this,
          i: a
        }) : ec(this.b.A, {
          f: this,
          c: a
        });
      }

      var c = this.ba(a);
      if (!c) return this.S(a), null;
      var d = dc(this.b, c);

      if (void 0 !== d) {
        if (0 === d.a.count.value) return d.a.c = c, d.a.i = a, d.clone();
        d = d.clone();
        this.S(a);
        return d;
      }

      d = this.b.aa(c);
      d = Mb[d];
      if (!d) return b.call(this);
      d = this.I ? d.Y : d.pointerType;
      var f = Xb(c, this.b, d.b);
      return null === f ? b.call(this) : this.J ? ec(d.b.A, {
        f: d,
        c: f,
        l: this,
        i: a
      }) : ec(d.b.A, {
        f: d,
        c: f
      });
    }

    function hc() {
      W.prototype.ba = Ub;
      W.prototype.S = Vb;
      W.prototype.argPackAdvance = 8;
      W.prototype.readValueFromPointer = L;
      W.prototype.deleteObject = Wb;
      W.prototype.fromWireType = fc;
    }

    function W(a, b, c, d, f, g, h, l, n, m, p) {
      this.name = a;
      this.b = b;
      this.N = c;
      this.I = d;
      this.J = f;
      this.la = g;
      this.ra = h;
      this.X = l;
      this.O = n;
      this.na = m;
      this.s = p;
      f || void 0 !== b.o ? this.toWireType = Sb : (this.toWireType = d ? Rb : Tb, this.j = null);
    }

    function ic(a, b, c) {
      e.hasOwnProperty(a) || rb("Replacing nonexistant public symbol");
      void 0 !== e[a].g && void 0 !== c ? e[a].g[c] = b : (e[a] = b, e[a].H = c);
    }

    function X(a, b) {
      a = R(a);
      if (void 0 !== e["FUNCTION_TABLE_" + a]) var c = e["FUNCTION_TABLE_" + a][b];else if ("undefined" !== typeof FUNCTION_TABLE) c = FUNCTION_TABLE[b];else {
        c = e["dynCall_" + a];
        void 0 === c && (c = e["dynCall_" + a.replace(/f/g, "d")], void 0 === c && S("No dynCall invoker for signature: " + a));

        for (var d = [], f = 1; f < a.length; ++f) d.push("a" + f);

        f = "return function " + ("dynCall_" + a + "_" + b) + "(" + d.join(", ") + ") {\n";
        f += "    return dynCall(rawFunction" + (d.length ? ", " : "") + d.join(", ") + ");\n";
        c = new Function("dynCall", "rawFunction", f + "};\n")(c, b);
      }
      "function" !== typeof c && S("unknown function pointer with signature " + a + ": " + b);
      return c;
    }

    var jc = void 0;

    function kc(a) {
      a = lc(a);
      var b = R(a);
      K(a);
      return b;
    }

    function mc(a, b) {
      function c(a) {
        f[a] || N[a] || (mb[a] ? mb[a].forEach(c) : (d.push(a), f[a] = !0));
      }

      var d = [],
          f = {};
      b.forEach(c);
      throw new jc(a + ": " + d.map(kc).join([", "]));
    }

    function nc(a, b) {
      for (var c = [], d = 0; d < a; d++) c.push(x[(b >> 2) + d]);

      return c;
    }

    function oc(a, b) {
      if (!(a instanceof Function)) throw new TypeError("new_ called with constructor type " + typeof a + " which is not a function");
      var c = ob(a.name || "unknownFunctionName", function () {});
      c.prototype = a.prototype;
      c = new c();
      a = a.apply(c, b);
      return a instanceof Object ? a : c;
    }

    function pc(a, b, c, d, f) {
      var g = b.length;
      2 > g && S("argTypes array size mismatch! Must at least get return value and 'this' types!");
      var h = null !== b[1] && null !== c,
          l = !1;

      for (c = 1; c < b.length; ++c) if (null !== b[c] && void 0 === b[c].j) {
        l = !0;
        break;
      }

      var n = "void" !== b[0].name,
          m = "",
          p = "";

      for (c = 0; c < g - 2; ++c) m += (0 !== c ? ", " : "") + "arg" + c, p += (0 !== c ? ", " : "") + "arg" + c + "Wired";

      a = "return function " + nb(a) + "(" + m + ") {\nif (arguments.length !== " + (g - 2) + ") {\nthrowBindingError('function " + a + " called with ' + arguments.length + ' arguments, expected " + (g - 2) + " args!');\n}\n";
      l && (a += "var destructors = [];\n");
      var w = l ? "destructors" : "null";
      m = "throwBindingError invoker fn runDestructors retType classParam".split(" ");
      d = [S, d, f, lb, b[0], b[1]];
      h && (a += "var thisWired = classParam.toWireType(" + w + ", this);\n");

      for (c = 0; c < g - 2; ++c) a += "var arg" + c + "Wired = argType" + c + ".toWireType(" + w + ", arg" + c + "); // " + b[c + 2].name + "\n", m.push("argType" + c), d.push(b[c + 2]);

      h && (p = "thisWired" + (0 < p.length ? ", " : "") + p);
      a += (n ? "var rv = " : "") + "invoker(fn" + (0 < p.length ? ", " : "") + p + ");\n";
      if (l) a += "runDestructors(destructors);\n";else for (c = h ? 1 : 2; c < b.length; ++c) g = 1 === c ? "thisWired" : "arg" + (c - 2) + "Wired", null !== b[c].j && (a += g + "_dtor(" + g + "); // " + b[c].name + "\n", m.push(g + "_dtor"), d.push(b[c].j));
      n && (a += "var ret = retType.fromWireType(rv);\nreturn ret;\n");
      m.push(a + "}\n");
      return oc(Function, m).apply(null, d);
    }

    var qc = [],
        Y = [{}, {
      value: void 0
    }, {
      value: null
    }, {
      value: !0
    }, {
      value: !1
    }];

    function rc(a) {
      4 < a && 0 === --Y[a].P && (Y[a] = void 0, qc.push(a));
    }

    function sc() {
      for (var a = 0, b = 5; b < Y.length; ++b) void 0 !== Y[b] && ++a;

      return a;
    }

    function tc() {
      for (var a = 5; a < Y.length; ++a) if (void 0 !== Y[a]) return Y[a];

      return null;
    }

    function uc() {
      e.count_emval_handles = sc;
      e.get_first_emval = tc;
    }

    function V(a) {
      switch (a) {
        case void 0:
          return 1;

        case null:
          return 2;

        case !0:
          return 3;

        case !1:
          return 4;

        default:
          var b = qc.length ? qc.pop() : Y.length;
          Y[b] = {
            P: 1,
            value: a
          };
          return b;
      }
    }

    function vc(a, b, c) {
      switch (b) {
        case 0:
          return function (a) {
            return this.fromWireType((c ? B : A)[a]);
          };

        case 1:
          return function (a) {
            return this.fromWireType((c ? Da : Ea)[a >> 1]);
          };

        case 2:
          return function (a) {
            return this.fromWireType((c ? x : C)[a >> 2]);
          };

        default:
          throw new TypeError("Unknown integer type: " + a);
      }
    }

    function wc(a, b) {
      var c = N[a];
      void 0 === c && S(b + " has unknown type " + kc(a));
      return c;
    }

    function U(a) {
      if (null === a) return "null";
      var b = typeof a;
      return "object" === b || "array" === b || "function" === b ? a.toString() : "" + a;
    }

    function xc(a, b) {
      switch (b) {
        case 2:
          return function (a) {
            return this.fromWireType(Fa[a >> 2]);
          };

        case 3:
          return function (a) {
            return this.fromWireType(Ga[a >> 3]);
          };

        default:
          throw new TypeError("Unknown float type: " + a);
      }
    }

    function yc(a, b, c) {
      switch (b) {
        case 0:
          return c ? function (a) {
            return B[a];
          } : function (a) {
            return A[a];
          };

        case 1:
          return c ? function (a) {
            return Da[a >> 1];
          } : function (a) {
            return Ea[a >> 1];
          };

        case 2:
          return c ? function (a) {
            return x[a >> 2];
          } : function (a) {
            return C[a >> 2];
          };

        default:
          throw new TypeError("Unknown integer type: " + a);
      }
    }

    function Z(a) {
      a || S("Cannot use deleted val. handle = " + a);
      return Y[a].value;
    }

    function zc(a, b) {
      for (var c = Array(a), d = 0; d < a; ++d) c[d] = wc(x[(b >> 2) + d], "parameter " + d);

      return c;
    }

    function Ac(a) {
      var b = [];
      x[a >> 2] = V(b);
      return b;
    }

    var Bc = {};

    function Cc(a) {
      var b = Bc[a];
      return void 0 === b ? R(a) : b;
    }

    var Dc = [];

    function Ec() {
      return "object" === typeof globalThis ? globalThis : Function("return this")();
    }

    function Fc(a) {
      var b = Dc.length;
      Dc.push(a);
      return b;
    }

    function Gc(a) {
      for (var b = "", c = 0; c < a; ++c) b += (0 !== c ? ", " : "") + "arg" + c;

      var d = "return function emval_allocator_" + a + "(constructor, argTypes, args) {\n";

      for (c = 0; c < a; ++c) d += "var argType" + c + " = requireRegisteredType(Module['HEAP32'][(argTypes >> 2) + " + c + '], "parameter ' + c + '");\nvar arg' + c + " = argType" + c + ".readValueFromPointer(args);\nargs += argType" + c + "['argPackAdvance'];\n";

      return new Function("requireRegisteredType", "Module", "__emval_register", d + ("var obj = new constructor(" + b + ");\nreturn __emval_register(obj);\n}\n"))(wc, e, V);
    }

    var Hc = {};

    function Ic() {
      u();
    }

    function Jc() {
      return ea || "undefined" !== typeof dateNow || "object" === typeof performance && performance && "function" === typeof performance.now;
    }

    function ra() {
      return B.length;
    }

    function Kc(a, b, c) {
      A.set(A.subarray(b, b + c), a);
    }

    function Lc(a) {
      try {
        return y.grow(a - buffer.byteLength + 65535 >> 16), Ha(y.buffer), 1;
      } catch (b) {}
    }

    function Mc(a) {
      if (0 === a) return 0;
      a = z(a);
      if (!F.hasOwnProperty(a)) return 0;
      Mc.m && K(Mc.m);
      a = F[a];
      var b = za(a) + 1,
          c = G(b);
      c && ya(a, B, c, b);
      Mc.m = c;
      return Mc.m;
    }

    function Nc(a) {
      a /= 1E3;
      if ((q || r) && self.performance && self.performance.now) for (var b = self.performance.now(); self.performance.now() - b < a;);else for (b = Date.now(); Date.now() - b < a;);
      return 0;
    }

    function Oc(a) {
      return 0 === a % 4 && (0 !== a % 100 || 0 === a % 400);
    }

    function Pc(a, b) {
      for (var c = 0, d = 0; d <= b; c += a[d++]);

      return c;
    }

    var Qc = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31],
        Rc = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

    function Sc(a, b) {
      for (a = new Date(a.getTime()); 0 < b;) {
        var c = a.getMonth(),
            d = (Oc(a.getFullYear()) ? Qc : Rc)[c];
        if (b > d - a.getDate()) b -= d - a.getDate() + 1, a.setDate(1), 11 > c ? a.setMonth(c + 1) : (a.setMonth(0), a.setFullYear(a.getFullYear() + 1));else {
          a.setDate(a.getDate() + b);
          break;
        }
      }

      return a;
    }

    function Tc(a, b, c, d) {
      function f(a, b, c) {
        for (a = "number" === typeof a ? a.toString() : a || ""; a.length < b;) a = c[0] + a;

        return a;
      }

      function g(a, b) {
        return f(a, b, "0");
      }

      function h(a, b) {
        function c(a) {
          return 0 > a ? -1 : 0 < a ? 1 : 0;
        }

        var d;
        0 === (d = c(a.getFullYear() - b.getFullYear())) && 0 === (d = c(a.getMonth() - b.getMonth())) && (d = c(a.getDate() - b.getDate()));
        return d;
      }

      function l(a) {
        switch (a.getDay()) {
          case 0:
            return new Date(a.getFullYear() - 1, 11, 29);

          case 1:
            return a;

          case 2:
            return new Date(a.getFullYear(), 0, 3);

          case 3:
            return new Date(a.getFullYear(), 0, 2);

          case 4:
            return new Date(a.getFullYear(), 0, 1);

          case 5:
            return new Date(a.getFullYear() - 1, 11, 31);

          case 6:
            return new Date(a.getFullYear() - 1, 11, 30);
        }
      }

      function n(a) {
        a = Sc(new Date(a.h + 1900, 0, 1), a.M);
        var b = l(new Date(a.getFullYear() + 1, 0, 4));
        return 0 >= h(l(new Date(a.getFullYear(), 0, 4)), a) ? 0 >= h(b, a) ? a.getFullYear() + 1 : a.getFullYear() : a.getFullYear() - 1;
      }

      var m = x[d + 40 >> 2];
      d = {
        ua: x[d >> 2],
        ta: x[d + 4 >> 2],
        K: x[d + 8 >> 2],
        D: x[d + 12 >> 2],
        B: x[d + 16 >> 2],
        h: x[d + 20 >> 2],
        L: x[d + 24 >> 2],
        M: x[d + 28 >> 2],
        Ca: x[d + 32 >> 2],
        sa: x[d + 36 >> 2],
        va: m ? z(m) : ""
      };
      c = z(c);
      m = {
        "%c": "%a %b %d %H:%M:%S %Y",
        "%D": "%m/%d/%y",
        "%F": "%Y-%m-%d",
        "%h": "%b",
        "%r": "%I:%M:%S %p",
        "%R": "%H:%M",
        "%T": "%H:%M:%S",
        "%x": "%m/%d/%y",
        "%X": "%H:%M:%S",
        "%Ec": "%c",
        "%EC": "%C",
        "%Ex": "%m/%d/%y",
        "%EX": "%H:%M:%S",
        "%Ey": "%y",
        "%EY": "%Y",
        "%Od": "%d",
        "%Oe": "%e",
        "%OH": "%H",
        "%OI": "%I",
        "%Om": "%m",
        "%OM": "%M",
        "%OS": "%S",
        "%Ou": "%u",
        "%OU": "%U",
        "%OV": "%V",
        "%Ow": "%w",
        "%OW": "%W",
        "%Oy": "%y"
      };

      for (var p in m) c = c.replace(new RegExp(p, "g"), m[p]);

      var w = "Sunday Monday Tuesday Wednesday Thursday Friday Saturday".split(" "),
          fa = "January February March April May June July August September October November December".split(" ");
      m = {
        "%a": function a(_a) {
          return w[_a.L].substring(0, 3);
        },
        "%A": function A(a) {
          return w[a.L];
        },
        "%b": function b(a) {
          return fa[a.B].substring(0, 3);
        },
        "%B": function B(a) {
          return fa[a.B];
        },
        "%C": function C(a) {
          return g((a.h + 1900) / 100 | 0, 2);
        },
        "%d": function d(a) {
          return g(a.D, 2);
        },
        "%e": function e(a) {
          return f(a.D, 2, " ");
        },
        "%g": function g(a) {
          return n(a).toString().substring(2);
        },
        "%G": function G(a) {
          return n(a);
        },
        "%H": function H(a) {
          return g(a.K, 2);
        },
        "%I": function I(a) {
          a = a.K;
          0 == a ? a = 12 : 12 < a && (a -= 12);
          return g(a, 2);
        },
        "%j": function j(a) {
          return g(a.D + Pc(Oc(a.h + 1900) ? Qc : Rc, a.B - 1), 3);
        },
        "%m": function m(a) {
          return g(a.B + 1, 2);
        },
        "%M": function M(a) {
          return g(a.ta, 2);
        },
        "%n": function n() {
          return "\n";
        },
        "%p": function p(a) {
          return 0 <= a.K && 12 > a.K ? "AM" : "PM";
        },
        "%S": function S(a) {
          return g(a.ua, 2);
        },
        "%t": function t() {
          return "\t";
        },
        "%u": function u(a) {
          return a.L || 7;
        },
        "%U": function U(a) {
          var b = new Date(a.h + 1900, 0, 1),
              c = 0 === b.getDay() ? b : Sc(b, 7 - b.getDay());
          a = new Date(a.h + 1900, a.B, a.D);
          return 0 > h(c, a) ? g(Math.ceil((31 - c.getDate() + (Pc(Oc(a.getFullYear()) ? Qc : Rc, a.getMonth() - 1) - 31) + a.getDate()) / 7), 2) : 0 === h(c, b) ? "01" : "00";
        },
        "%V": function V(a) {
          var b = l(new Date(a.h + 1900, 0, 4)),
              c = l(new Date(a.h + 1901, 0, 4)),
              d = Sc(new Date(a.h + 1900, 0, 1), a.M);
          return 0 > h(d, b) ? "53" : 0 >= h(c, d) ? "01" : g(Math.ceil((b.getFullYear() < a.h + 1900 ? a.M + 32 - b.getDate() : a.M + 1 - b.getDate()) / 7), 2);
        },
        "%w": function w(a) {
          return a.L;
        },
        "%W": function W(a) {
          var b = new Date(a.h, 0, 1),
              c = 1 === b.getDay() ? b : Sc(b, 0 === b.getDay() ? 1 : 7 - b.getDay() + 1);
          a = new Date(a.h + 1900, a.B, a.D);
          return 0 > h(c, a) ? g(Math.ceil((31 - c.getDate() + (Pc(Oc(a.getFullYear()) ? Qc : Rc, a.getMonth() - 1) - 31) + a.getDate()) / 7), 2) : 0 === h(c, b) ? "01" : "00";
        },
        "%y": function y(a) {
          return (a.h + 1900).toString().substring(2);
        },
        "%Y": function Y(a) {
          return a.h + 1900;
        },
        "%z": function z(a) {
          a = a.sa;
          var b = 0 <= a;
          a = Math.abs(a) / 60;
          return (b ? "+" : "-") + String("0000" + (a / 60 * 100 + a % 60)).slice(-4);
        },
        "%Z": function Z(a) {
          return a.va;
        },
        "%%": function _() {
          return "%";
        }
      };

      for (p in m) 0 <= c.indexOf(p) && (c = c.replace(new RegExp(p, "g"), m[p](d)));

      p = Uc(c);
      if (p.length > b) return 0;
      B.set(p, a);
      return p.length - 1;
    }

    qb = e.InternalError = pb(Error, "InternalError");
    tb();
    vb = e.BindingError = pb(Error, "BindingError");
    Lb();
    hc();
    bc();
    jc = e.UnboundTypeError = pb(Error, "UnboundTypeError");
    uc();
    ea ? Ic = function Ic() {
      var a = process.hrtime();
      return 1E3 * a[0] + a[1] / 1E6;
    } : "undefined" !== typeof dateNow ? Ic = dateNow : "object" === typeof performance && performance && "function" === typeof performance.now ? Ic = function Ic() {
      return performance.now();
    } : Ic = Date.now;

    function Uc(a) {
      var b = Array(za(a) + 1);
      ya(a, b, 0, b.length);
      return b;
    }

    var Wc = {
      ClassHandle: T,
      ClassHandle_clone: Eb,
      ClassHandle_delete: Fb,
      ClassHandle_deleteLater: Kb,
      ClassHandle_isAliasOf: wb,
      ClassHandle_isDeleted: Gb,
      RegisteredClass: Pb,
      RegisteredPointer: W,
      RegisteredPointer_deleteObject: Wb,
      RegisteredPointer_destructor: Vb,
      RegisteredPointer_fromWireType: fc,
      RegisteredPointer_getPointee: Ub,
      __buildEnvironment: ab,
      __cxa_allocate_exception: function __cxa_allocate_exception(a) {
        return G(a);
      },
      __cxa_atexit: function __cxa_atexit() {
        return bb.apply(null, arguments);
      },
      __cxa_pure_virtual: function __cxa_pure_virtual() {
        va = !0;
        throw "Pure virtual function called!";
      },
      __cxa_throw: function __cxa_throw(a) {
        "uncaught_exception" in Vc ? Vc.m++ : Vc.m = 1;
        throw a;
      },
      __cxa_uncaught_exceptions: function __cxa_uncaught_exceptions() {
        return Vc.m;
      },
      __lock: function __lock() {},
      __map_file: function __map_file() {
        cb(1);
        return -1;
      },
      __setErrNo: cb,
      __syscall10: function __syscall10(a, b) {
        H = b;

        try {
          var c = z(I());
          (void 0).unlink(c);
          return 0;
        } catch (d) {
          return u(d), -d.u;
        }
      },
      __syscall140: function __syscall140(a, b) {
        H = b;

        try {
          return J.V(), I(), I(), I(), I(), 0;
        } catch (c) {
          return u(c), -c.u;
        }
      },
      __syscall145: function __syscall145(a, b) {
        H = b;

        try {
          var c = J.V(),
              d = I(),
              f = I();
          return J.xa(c, d, f);
        } catch (g) {
          return u(g), -g.u;
        }
      },
      __syscall163: function __syscall163(a, b) {
        H = b;
        return -12;
      },
      __syscall192: function __syscall192(a, b) {
        H = b;

        try {
          var c = I(),
              d = I(),
              f = I(),
              g = I(),
              h = I(),
              l = I();
          return gb(c, d, f, g, h, l);
        } catch (n) {
          return u(n), -n.u;
        }
      },
      __syscall194: function __syscall194(a, b) {
        H = b;

        try {
          var c = I();
          I();
          var d = I();
          I();
          (void 0).ya(c, d);
          return 0;
        } catch (f) {
          return u(f), -f.u;
        }
      },
      __syscall221: function __syscall221(a, b) {
        H = b;
        return 0;
      },
      __syscall5: function __syscall5(a, b) {
        H = b;

        try {
          var c = z(I()),
              d = I(),
              f = I();
          return (void 0).open(c, d, f).fd;
        } catch (g) {
          return u(g), -g.u;
        }
      },
      __syscall54: function __syscall54(a, b) {
        H = b;
        return 0;
      },
      __syscall6: function __syscall6(a, b) {
        H = b;

        try {
          return J.V(), 0;
        } catch (c) {
          return u(c), -c.u;
        }
      },
      __syscall91: function __syscall91(a, b) {
        H = b;

        try {
          var c = I(),
              d = I();
          return ib(c, d);
        } catch (f) {
          return u(f), -f.u;
        }
      },
      __unlock: function __unlock() {},
      __wasi_fd_write: function __wasi_fd_write() {
        return jb.apply(null, arguments);
      },
      _addDays: Sc,
      _arraySum: Pc,
      _embind_finalize_value_object: function _embind_finalize_value_object(a) {
        var b = kb[a];
        delete kb[a];
        var c = b.O,
            d = b.s,
            f = b.T,
            g = f.map(function (a) {
          return a.ga;
        }).concat(f.map(function (a) {
          return a.pa;
        }));
        O([a], g, function (a) {
          var g = {};
          f.forEach(function (b, c) {
            var d = a[c],
                l = b.ea,
                h = b.fa,
                m = a[c + f.length],
                n = b.oa,
                $c = b.qa;
            g[b.$] = {
              read: function read(a) {
                return d.fromWireType(l(h, a));
              },
              write: function write(a, b) {
                var c = [];
                n($c, a, m.toWireType(c, b));
                lb(c);
              }
            };
          });
          return [{
            name: b.name,
            fromWireType: function fromWireType(a) {
              var b = {},
                  c;

              for (c in g) b[c] = g[c].read(a);

              d(a);
              return b;
            },
            toWireType: function toWireType(a, b) {
              for (var f in g) if (!(f in b)) throw new TypeError("Missing field");

              var l = c();

              for (f in g) g[f].write(l, b[f]);

              null !== a && a.push(d, l);
              return l;
            },
            argPackAdvance: 8,
            readValueFromPointer: L,
            j: d
          }];
        });
      },
      _embind_register_bool: function _embind_register_bool(a, b, c, d, f) {
        var g = sb(c);
        b = R(b);
        Q(a, {
          name: b,
          fromWireType: function fromWireType(a) {
            return !!a;
          },
          toWireType: function toWireType(a, b) {
            return b ? d : f;
          },
          argPackAdvance: 8,
          readValueFromPointer: function readValueFromPointer(a) {
            if (1 === c) var d = B;else if (2 === c) d = Da;else if (4 === c) d = x;else throw new TypeError("Unknown boolean type size: " + b);
            return this.fromWireType(d[a >> g]);
          },
          j: null
        });
      },
      _embind_register_class: function _embind_register_class(a, b, c, d, f, g, h, l, n, m, p, w, fa) {
        p = R(p);
        g = X(f, g);
        l && (l = X(h, l));
        m && (m = X(n, m));
        fa = X(w, fa);
        var P = nb(p);
        Ob(P, function () {
          mc("Cannot construct " + p + " due to unbound types", [d]);
        });
        O([a, b, c], d ? [d] : [], function (b) {
          b = b[0];

          if (d) {
            var c = b.b;
            var f = c.A;
          } else f = T.prototype;

          b = ob(P, function () {
            if (Object.getPrototypeOf(this) !== h) throw new vb("Use 'new' to construct " + p);
            if (void 0 === n.v) throw new vb(p + " has no accessible constructor");
            var a = n.v[arguments.length];
            if (void 0 === a) throw new vb("Tried to invoke ctor of " + p + " with invalid number of parameters (" + arguments.length + ") - expected (" + Object.keys(n.v).toString() + ") parameters instead!");
            return a.apply(this, arguments);
          });
          var h = Object.create(f, {
            constructor: {
              value: b
            }
          });
          b.prototype = h;
          var n = new Pb(p, b, h, fa, c, g, l, m);
          c = new W(p, n, !0, !1, !1);
          f = new W(p + "*", n, !1, !1, !1);
          var w = new W(p + " const*", n, !1, !0, !1);
          Mb[a] = {
            pointerType: f,
            Y: w
          };
          ic(P, b);
          return [c, f, w];
        });
      },
      _embind_register_class_constructor: function _embind_register_class_constructor(a, b, c, d, f, g) {
        var h = nc(b, c);
        f = X(d, f);
        O([], [a], function (a) {
          a = a[0];
          var c = "constructor " + a.name;
          void 0 === a.b.v && (a.b.v = []);
          if (void 0 !== a.b.v[b - 1]) throw new vb("Cannot register multiple constructors with identical number of parameters (" + (b - 1) + ") for class '" + a.name + "'! Overload resolution is currently only performed using the parameter count, not actual type info!");

          a.b.v[b - 1] = function () {
            mc("Cannot construct " + a.name + " due to unbound types", h);
          };

          O([], h, function (d) {
            a.b.v[b - 1] = function () {
              arguments.length !== b - 1 && S(c + " called with " + arguments.length + " arguments, expected " + (b - 1));
              var a = [],
                  h = Array(b);
              h[0] = g;

              for (var l = 1; l < b; ++l) h[l] = d[l].toWireType(a, arguments[l - 1]);

              h = f.apply(null, h);
              lb(a);
              return d[0].fromWireType(h);
            };

            return [];
          });
          return [];
        });
      },
      _embind_register_class_function: function _embind_register_class_function(a, b, c, d, f, g, h, l) {
        var n = nc(c, d);
        b = R(b);
        g = X(f, g);
        O([], [a], function (a) {
          function d() {
            mc("Cannot call " + f + " due to unbound types", n);
          }

          a = a[0];
          var f = a.name + "." + b;
          l && a.b.ma.push(b);
          var m = a.b.A,
              P = m[b];
          void 0 === P || void 0 === P.g && P.className !== a.name && P.H === c - 2 ? (d.H = c - 2, d.className = a.name, m[b] = d) : (Nb(m, b, f), m[b].g[c - 2] = d);
          O([], n, function (d) {
            d = pc(f, d, a, g, h);
            void 0 === m[b].g ? (d.H = c - 2, m[b] = d) : m[b].g[c - 2] = d;
            return [];
          });
          return [];
        });
      },
      _embind_register_emval: function _embind_register_emval(a, b) {
        b = R(b);
        Q(a, {
          name: b,
          fromWireType: function fromWireType(a) {
            var b = Y[a].value;
            rc(a);
            return b;
          },
          toWireType: function toWireType(a, b) {
            return V(b);
          },
          argPackAdvance: 8,
          readValueFromPointer: L,
          j: null
        });
      },
      _embind_register_enum: function _embind_register_enum(a, b, c, d) {
        function f() {}

        c = sb(c);
        b = R(b);
        f.values = {};
        Q(a, {
          name: b,
          constructor: f,
          fromWireType: function fromWireType(a) {
            return this.constructor.values[a];
          },
          toWireType: function toWireType(a, b) {
            return b.value;
          },
          argPackAdvance: 8,
          readValueFromPointer: vc(b, c, d),
          j: null
        });
        Ob(b, f);
      },
      _embind_register_enum_value: function _embind_register_enum_value(a, b, c) {
        var d = wc(a, "enum");
        b = R(b);
        a = d.constructor;
        d = Object.create(d.constructor.prototype, {
          value: {
            value: c
          },
          constructor: {
            value: ob(d.name + "_" + b, function () {})
          }
        });
        a.values[c] = d;
        a[b] = d;
      },
      _embind_register_float: function _embind_register_float(a, b, c) {
        c = sb(c);
        b = R(b);
        Q(a, {
          name: b,
          fromWireType: function fromWireType(a) {
            return a;
          },
          toWireType: function toWireType(a, b) {
            if ("number" !== typeof b && "boolean" !== typeof b) throw new TypeError('Cannot convert "' + U(b) + '" to ' + this.name);
            return b;
          },
          argPackAdvance: 8,
          readValueFromPointer: xc(b, c),
          j: null
        });
      },
      _embind_register_function: function _embind_register_function(a, b, c, d, f, g) {
        var h = nc(b, c);
        a = R(a);
        f = X(d, f);
        Ob(a, function () {
          mc("Cannot call " + a + " due to unbound types", h);
        }, b - 1);
        O([], h, function (c) {
          c = [c[0], null].concat(c.slice(1));
          ic(a, pc(a, c, null, f, g), b - 1);
          return [];
        });
      },
      _embind_register_integer: function _embind_register_integer(a, b, c, d, f) {
        function g(a) {
          return a;
        }

        b = R(b);
        -1 === f && (f = 4294967295);
        var h = sb(c);

        if (0 === d) {
          var l = 32 - 8 * c;

          g = function g(a) {
            return a << l >>> l;
          };
        }

        var n = -1 != b.indexOf("unsigned");
        Q(a, {
          name: b,
          fromWireType: g,
          toWireType: function toWireType(a, c) {
            if ("number" !== typeof c && "boolean" !== typeof c) throw new TypeError('Cannot convert "' + U(c) + '" to ' + this.name);
            if (c < d || c > f) throw new TypeError('Passing a number "' + U(c) + '" from JS side to C/C++ side to an argument of type "' + b + '", which is outside the valid range [' + d + ", " + f + "]!");
            return n ? c >>> 0 : c | 0;
          },
          argPackAdvance: 8,
          readValueFromPointer: yc(b, h, 0 !== d),
          j: null
        });
      },
      _embind_register_memory_view: function _embind_register_memory_view(a, b, c) {
        function d(a) {
          a >>= 2;
          var b = C;
          return new f(b.buffer, b[a + 1], b[a]);
        }

        var f = [Int8Array, Uint8Array, Int16Array, Uint16Array, Int32Array, Uint32Array, Float32Array, Float64Array][b];
        c = R(c);
        Q(a, {
          name: c,
          fromWireType: d,
          argPackAdvance: 8,
          readValueFromPointer: d
        }, {
          ha: !0
        });
      },
      _embind_register_smart_ptr: function _embind_register_smart_ptr(a, b, c, d, f, g, h, l, n, m, p, w) {
        c = R(c);
        g = X(f, g);
        l = X(h, l);
        m = X(n, m);
        w = X(p, w);
        O([a], [b], function (a) {
          a = a[0];
          return [new W(c, a.b, !1, !1, !0, a, d, g, l, m, w)];
        });
      },
      _embind_register_std_string: function _embind_register_std_string(a, b) {
        b = R(b);
        var c = "std::string" === b;
        Q(a, {
          name: b,
          fromWireType: function fromWireType(a) {
            var b = C[a >> 2];

            if (c) {
              var d = A[a + 4 + b],
                  h = 0;
              0 != d && (h = d, A[a + 4 + b] = 0);
              var l = a + 4;

              for (d = 0; d <= b; ++d) {
                var n = a + 4 + d;

                if (0 == A[n]) {
                  l = z(l);
                  if (void 0 === m) var m = l;else m += String.fromCharCode(0), m += l;
                  l = n + 1;
                }
              }

              0 != h && (A[a + 4 + b] = h);
            } else {
              m = Array(b);

              for (d = 0; d < b; ++d) m[d] = String.fromCharCode(A[a + 4 + d]);

              m = m.join("");
            }

            K(a);
            return m;
          },
          toWireType: function toWireType(a, b) {
            b instanceof ArrayBuffer && (b = new Uint8Array(b));
            var d = "string" === typeof b;
            d || b instanceof Uint8Array || b instanceof Uint8ClampedArray || b instanceof Int8Array || S("Cannot pass non-string to std::string");
            var f = (c && d ? function () {
              return za(b);
            } : function () {
              return b.length;
            })(),
                l = G(4 + f + 1);
            C[l >> 2] = f;
            if (c && d) ya(b, A, l + 4, f + 1);else if (d) for (d = 0; d < f; ++d) {
              var n = b.charCodeAt(d);
              255 < n && (K(l), S("String has UTF-16 code units that do not fit in 8 bits"));
              A[l + 4 + d] = n;
            } else for (d = 0; d < f; ++d) A[l + 4 + d] = b[d];
            null !== a && a.push(K, l);
            return l;
          },
          argPackAdvance: 8,
          readValueFromPointer: L,
          j: function j(a) {
            K(a);
          }
        });
      },
      _embind_register_std_wstring: function _embind_register_std_wstring(a, b, c) {
        c = R(c);

        if (2 === b) {
          var d = function d() {
            return Ea;
          };

          var f = 1;
        } else 4 === b && (d = function d() {
          return C;
        }, f = 2);

        Q(a, {
          name: c,
          fromWireType: function fromWireType(a) {
            for (var b = d(), c = C[a >> 2], g = Array(c), m = a + 4 >> f, p = 0; p < c; ++p) g[p] = String.fromCharCode(b[m + p]);

            K(a);
            return g.join("");
          },
          toWireType: function toWireType(a, c) {
            var g = d(),
                h = c.length,
                m = G(4 + h * b);
            C[m >> 2] = h;

            for (var p = m + 4 >> f, w = 0; w < h; ++w) g[p + w] = c.charCodeAt(w);

            null !== a && a.push(K, m);
            return m;
          },
          argPackAdvance: 8,
          readValueFromPointer: L,
          j: function j(a) {
            K(a);
          }
        });
      },
      _embind_register_value_object: function _embind_register_value_object(a, b, c, d, f, g) {
        kb[a] = {
          name: R(b),
          O: X(c, d),
          s: X(f, g),
          T: []
        };
      },
      _embind_register_value_object_field: function _embind_register_value_object_field(a, b, c, d, f, g, h, l, n, m) {
        kb[a].T.push({
          $: R(b),
          ga: c,
          ea: X(d, f),
          fa: g,
          pa: h,
          oa: X(l, n),
          qa: m
        });
      },
      _embind_register_void: function _embind_register_void(a, b) {
        b = R(b);
        Q(a, {
          ia: !0,
          name: b,
          argPackAdvance: 0,
          fromWireType: function fromWireType() {},
          toWireType: function toWireType() {}
        });
      },
      _emscripten_syscall_mmap2: gb,
      _emscripten_syscall_munmap: ib,
      _emval_addMethodCaller: Fc,
      _emval_allocateDestructors: Ac,
      _emval_as: function _emval_as(a, b, c) {
        a = Z(a);
        b = wc(b, "emval::as");
        var d = [],
            f = V(d);
        x[c >> 2] = f;
        return b.toWireType(d, a);
      },
      _emval_call: function _emval_call(a, b, c, d) {
        a = Z(a);
        c = zc(b, c);

        for (var f = Array(b), g = 0; g < b; ++g) {
          var h = c[g];
          f[g] = h.readValueFromPointer(d);
          d += h.argPackAdvance;
        }

        a = a.apply(void 0, f);
        return V(a);
      },
      _emval_call_method: function _emval_call_method(a, b, c, d, f) {
        a = Dc[a];
        b = Z(b);
        c = Cc(c);
        return a(b, c, Ac(d), f);
      },
      _emval_call_void_method: function _emval_call_void_method(a, b, c, d) {
        a = Dc[a];
        b = Z(b);
        c = Cc(c);
        a(b, c, null, d);
      },
      _emval_decref: rc,
      _emval_get_global: function _emval_get_global(a) {
        if (0 === a) return V(Ec());
        a = Cc(a);
        return V(Ec()[a]);
      },
      _emval_get_method_caller: function _emval_get_method_caller(a, b) {
        b = zc(a, b);

        for (var c = b[0], d = c.name + "_$" + b.slice(1).map(function (a) {
          return a.name;
        }).join("_") + "$", f = ["retType"], g = [c], h = "", l = 0; l < a - 1; ++l) h += (0 !== l ? ", " : "") + "arg" + l, f.push("argType" + l), g.push(b[1 + l]);

        d = "return function " + nb("methodCaller_" + d) + "(handle, name, destructors, args) {\n";
        var n = 0;

        for (l = 0; l < a - 1; ++l) d += "    var arg" + l + " = argType" + l + ".readValueFromPointer(args" + (n ? "+" + n : "") + ");\n", n += b[l + 1].argPackAdvance;

        d += "    var rv = handle[name](" + h + ");\n";

        for (l = 0; l < a - 1; ++l) b[l + 1].deleteObject && (d += "    argType" + l + ".deleteObject(arg" + l + ");\n");

        c.ia || (d += "    return retType.toWireType(destructors, rv);\n");
        f.push(d + "};\n");
        a = oc(Function, f).apply(null, g);
        return Fc(a);
      },
      _emval_get_module_property: function _emval_get_module_property(a) {
        a = Cc(a);
        return V(e[a]);
      },
      _emval_get_property: function _emval_get_property(a, b) {
        a = Z(a);
        b = Z(b);
        return V(a[b]);
      },
      _emval_incref: function _emval_incref(a) {
        4 < a && (Y[a].P += 1);
      },
      _emval_instanceof: function _emval_instanceof(a, b) {
        a = Z(a);
        b = Z(b);
        return a instanceof b;
      },
      _emval_lookupTypes: zc,
      _emval_new: function _emval_new(a, b, c, d) {
        a = Z(a);
        var f = Hc[b];
        f || (f = Gc(b), Hc[b] = f);
        return f(a, c, d);
      },
      _emval_new_cstring: function _emval_new_cstring(a) {
        return V(Cc(a));
      },
      _emval_new_object: function _emval_new_object() {
        return V({});
      },
      _emval_register: V,
      _emval_run_destructors: function _emval_run_destructors(a) {
        lb(Y[a].value);
        rc(a);
      },
      _emval_set_property: function _emval_set_property(a, b, c) {
        a = Z(a);
        b = Z(b);
        c = Z(c);
        a[b] = c;
      },
      _emval_take_value: function _emval_take_value(a, b) {
        a = wc(a, "_emval_take_value");
        a = a.readValueFromPointer(b);
        return V(a);
      },
      _emval_typeof: function _emval_typeof(a) {
        a = Z(a);
        return V(typeof a);
      },
      _isLeapYear: Oc,
      abort: function abort() {
        u();
      },
      abortOnCannotGrowMemory: function abortOnCannotGrowMemory() {
        u("OOM");
      },
      atexit: bb,
      attachFinalizer: Db,
      clock_gettime: function clock_gettime(a, b) {
        if (0 === a) a = Date.now();else if (1 === a && Jc()) a = Ic();else return cb(22), -1;
        x[b >> 2] = a / 1E3 | 0;
        x[b + 4 >> 2] = a % 1E3 * 1E6 | 0;
        return 0;
      },
      constNoSmartPtrRawPointerToWireType: Rb,
      count_emval_handles: sc,
      craftEmvalAllocator: Gc,
      craftInvokerFunction: pc,
      createNamedFunction: ob,
      demangle: function demangle(a) {
        return a;
      },
      demangleAll: Za,
      detachFinalizer: Ab,
      downcastPointer: Xb,
      embind__requireFunction: X,
      embind_init_charCodes: tb,
      embind_repr: U,
      emscripten_asm_const_iii: function emscripten_asm_const_iii(a, b, c) {
        a: for (var d = "";;) {
          var f = A[b++ >> 0];

          if (!f) {
            b = d;
            break a;
          }

          d += String.fromCharCode(f);
        }

        d = [];

        for (f = 0; f < b.length; f++) {
          var g = b[f];
          "d" == g || "f" == g ? (c = c + 8 - 1 & -8, d.push(Ga[c >> 3]), c += 8) : "i" == g && (c = c + 4 - 1 & -4, d.push(x[c >> 2]), c += 4);
        }

        return Xa[a].apply(null, d);
      },
      emscripten_get_heap_size: ra,
      emscripten_get_now: Ic,
      emscripten_get_now_is_monotonic: Jc,
      emscripten_get_sbrk_ptr: function emscripten_get_sbrk_ptr() {
        return 95376;
      },
      emscripten_memcpy_big: Kc,
      emscripten_realloc_buffer: Lc,
      emscripten_resize_heap: function emscripten_resize_heap(a) {
        if (2147418112 < a) return !1;

        for (var b = Math.max(ra(), 16777216); b < a;) 536870912 >= b ? b = Ca(2 * b) : b = Math.min(Ca((3 * b + 2147483648) / 4), 2147418112);

        return Lc(b) ? !0 : !1;
      },
      emval_get_global: Ec,
      ensureOverloadTable: Nb,
      enumReadValueFromPointer: vc,
      exposePublicSymbol: Ob,
      extendError: pb,
      fd_write: jb,
      floatReadValueFromPointer: xc,
      flushPendingDeletes: Jb,
      flush_NO_FILESYSTEM: function flush_NO_FILESYSTEM() {
        var a = e._fflush;
        a && a(0);
        db[1].length && eb(1, 10);
        db[2].length && eb(2, 10);
      },
      genericPointerToWireType: Sb,
      getBasestPointer: cc,
      getInheritedInstance: dc,
      getInheritedInstanceCount: Yb,
      getLiveInheritedInstances: $b,
      getShiftFromSize: sb,
      getStringOrSymbol: Cc,
      getTypeName: kc,
      get_first_emval: tc,
      getenv: Mc,
      getpagesize: function getpagesize() {
        return 16384;
      },
      heap32VectorToArray: nc,
      init_ClassHandle: Lb,
      init_RegisteredPointer: hc,
      init_embind: bc,
      init_emval: uc,
      integerReadValueFromPointer: yc,
      jsStackTrace: $a,
      makeClassHandle: ec,
      makeLegalFunctionName: nb,
      memcpy: function memcpy(a, b, c) {
        a |= 0;
        b |= 0;
        c |= 0;
        var d;
        if (8192 <= (c | 0)) return Kc(a | 0, b | 0, c | 0) | 0, a | 0;
        var f = a | 0;
        var g = a + c | 0;

        if ((a & 3) == (b & 3)) {
          for (; a & 3;) {
            if (0 == (c | 0)) return f | 0;
            B[a >> 0] = B[b >> 0] | 0;
            a = a + 1 | 0;
            b = b + 1 | 0;
            c = c - 1 | 0;
          }

          c = g & -4 | 0;

          for (d = c - 64 | 0; (a | 0) <= (d | 0);) x[a >> 2] = x[b >> 2] | 0, x[a + 4 >> 2] = x[b + 4 >> 2] | 0, x[a + 8 >> 2] = x[b + 8 >> 2] | 0, x[a + 12 >> 2] = x[b + 12 >> 2] | 0, x[a + 16 >> 2] = x[b + 16 >> 2] | 0, x[a + 20 >> 2] = x[b + 20 >> 2] | 0, x[a + 24 >> 2] = x[b + 24 >> 2] | 0, x[a + 28 >> 2] = x[b + 28 >> 2] | 0, x[a + 32 >> 2] = x[b + 32 >> 2] | 0, x[a + 36 >> 2] = x[b + 36 >> 2] | 0, x[a + 40 >> 2] = x[b + 40 >> 2] | 0, x[a + 44 >> 2] = x[b + 44 >> 2] | 0, x[a + 48 >> 2] = x[b + 48 >> 2] | 0, x[a + 52 >> 2] = x[b + 52 >> 2] | 0, x[a + 56 >> 2] = x[b + 56 >> 2] | 0, x[a + 60 >> 2] = x[b + 60 >> 2] | 0, a = a + 64 | 0, b = b + 64 | 0;

          for (; (a | 0) < (c | 0);) x[a >> 2] = x[b >> 2] | 0, a = a + 4 | 0, b = b + 4 | 0;
        } else for (c = g - 4 | 0; (a | 0) < (c | 0);) B[a >> 0] = B[b >> 0] | 0, B[a + 1 >> 0] = B[b + 1 >> 0] | 0, B[a + 2 >> 0] = B[b + 2 >> 0] | 0, B[a + 3 >> 0] = B[b + 3 >> 0] | 0, a = a + 4 | 0, b = b + 4 | 0;

        for (; (a | 0) < (g | 0);) B[a >> 0] = B[b >> 0] | 0, a = a + 1 | 0, b = b + 1 | 0;

        return f | 0;
      },
      memory: y,
      memset: fb,
      nanosleep: function nanosleep(a, b) {
        if (0 === a) return cb(22), -1;
        var c = x[a >> 2];
        a = x[a + 4 >> 2];
        if (0 > a || 999999999 < a || 0 > c) return cb(22), -1;
        0 !== b && (x[b >> 2] = 0, x[b + 4 >> 2] = 0);
        return Nc(1E6 * c + a / 1E3);
      },
      new_: oc,
      nonConstNoSmartPtrRawPointerToWireType: Tb,
      pthread_cond_broadcast: function pthread_cond_broadcast() {
        return 0;
      },
      pthread_cond_wait: function pthread_cond_wait() {
        return 0;
      },
      readLatin1String: R,
      registerType: Q,
      releaseClassHandle: Cb,
      replacePublicSymbol: ic,
      requireHandle: Z,
      requireRegisteredType: wc,
      runDestructor: Bb,
      runDestructors: lb,
      sched_yield: function sched_yield() {
        return 0;
      },
      setDelayFunction: ac,
      setTempRet0: function setTempRet0() {},
      shallowCopyInternalPointer: xb,
      simpleReadValueFromPointer: L,
      stackTrace: function stackTrace() {
        var a = $a();
        e.extraStackTrace && (a += "\n" + e.extraStackTrace());
        return Za(a);
      },
      strftime: Tc,
      strftime_l: function strftime_l(a, b, c, d) {
        return Tc(a, b, c, d);
      },
      table: ua,
      throwBindingError: S,
      throwInstanceAlreadyDeleted: yb,
      throwInternalError: rb,
      throwUnboundTypeError: mc,
      upcastPointer: Qb,
      usleep: Nc,
      whenDependentTypesAreResolved: O
    },
        Xc = function () {
      function a(a) {
        e.asm = a.exports;
        D--;
        e.monitorRunDependencies && e.monitorRunDependencies(D);
        0 == D && (null !== Ra && (clearInterval(Ra), Ra = null), Sa && (a = Sa, Sa = null, a()));
      }

      function b(b) {
        a(b.instance);
      }

      function c(a) {
        return Wa().then(function (a) {
          return WebAssembly.instantiate(a, d);
        }).then(a, function (a) {
          v("failed to asynchronously prepare wasm: " + a);
          u(a);
        });
      }

      var d = {
        env: Wc,
        wasi_unstable: Wc
      };
      D++;
      e.monitorRunDependencies && e.monitorRunDependencies(D);
      if (e.instantiateWasm) try {
        return e.instantiateWasm(d, a);
      } catch (f) {
        return v("Module.instantiateWasm callback failed with error: " + f), !1;
      }

      (function () {
        if (sa || "function" !== typeof WebAssembly.instantiateStreaming || Ta() || "function" !== typeof fetch) return c(b);
        fetch(E, {
          credentials: "same-origin"
        }).then(function (a) {
          return WebAssembly.instantiateStreaming(a, d).then(b, function (a) {
            v("wasm streaming compile failed: " + a);
            v("falling back to ArrayBuffer instantiation");
            c(b);
          });
        });
      })();

      return {};
    }();

    e.asm = Xc;

    var Ya = e.___wasm_call_ctors = function () {
      return e.asm.__wasm_call_ctors.apply(null, arguments);
    },
        G = e._malloc = function () {
      return e.asm.malloc.apply(null, arguments);
    },
        K = e._free = function () {
      return e.asm.free.apply(null, arguments);
    };

    e._main = function () {
      return e.asm.main.apply(null, arguments);
    };

    var Vc = e.__ZSt18uncaught_exceptionv = function () {
      return e.asm._ZSt18uncaught_exceptionv.apply(null, arguments);
    };

    e._htonl = function () {
      return e.asm.htonl.apply(null, arguments);
    };

    e._htons = function () {
      return e.asm.htons.apply(null, arguments);
    };

    e._ntohs = function () {
      return e.asm.ntohs.apply(null, arguments);
    };

    e._setThrew = function () {
      return e.asm.setThrew.apply(null, arguments);
    };

    e.___cxa_can_catch = function () {
      return e.asm.__cxa_can_catch.apply(null, arguments);
    };

    e.___cxa_is_pointer_type = function () {
      return e.asm.__cxa_is_pointer_type.apply(null, arguments);
    };

    e.__get_environ = function () {
      return e.asm._get_environ.apply(null, arguments);
    };

    e.___embind_register_native_and_builtin_types = function () {
      return e.asm.__embind_register_native_and_builtin_types.apply(null, arguments);
    };

    var lc = e.___getTypeName = function () {
      return e.asm.__getTypeName.apply(null, arguments);
    },
        hb = e._memalign = function () {
      return e.asm.memalign.apply(null, arguments);
    };

    e._emscripten_builtin_free = function () {
      return e.asm.emscripten_builtin_free.apply(null, arguments);
    };

    e._emscripten_builtin_memalign = function () {
      return e.asm.emscripten_builtin_memalign.apply(null, arguments);
    };

    e.stackSave = function () {
      return e.asm.stackSave.apply(null, arguments);
    };

    var Ba = e.stackAlloc = function () {
      return e.asm.stackAlloc.apply(null, arguments);
    };

    e.stackRestore = function () {
      return e.asm.stackRestore.apply(null, arguments);
    };

    e.__growWasmMemory = function () {
      return e.asm.__growWasmMemory.apply(null, arguments);
    };

    e.dynCall_ii = function () {
      return e.asm.dynCall_ii.apply(null, arguments);
    };

    e.dynCall_vi = function () {
      return e.asm.dynCall_vi.apply(null, arguments);
    };

    e.dynCall_iii = function () {
      return e.asm.dynCall_iii.apply(null, arguments);
    };

    e.dynCall_viij = function () {
      return e.asm.dynCall_viij.apply(null, arguments);
    };

    e.dynCall_v = function () {
      return e.asm.dynCall_v.apply(null, arguments);
    };

    e.dynCall_vii = function () {
      return e.asm.dynCall_vii.apply(null, arguments);
    };

    e.dynCall_viii = function () {
      return e.asm.dynCall_viii.apply(null, arguments);
    };

    e.dynCall_viiji = function () {
      return e.asm.dynCall_viiji.apply(null, arguments);
    };

    e.dynCall_viijji = function () {
      return e.asm.dynCall_viijji.apply(null, arguments);
    };

    e.dynCall_ji = function () {
      return e.asm.dynCall_ji.apply(null, arguments);
    };

    e.dynCall_viiiiii = function () {
      return e.asm.dynCall_viiiiii.apply(null, arguments);
    };

    e.dynCall_viiii = function () {
      return e.asm.dynCall_viiii.apply(null, arguments);
    };

    e.dynCall_viijj = function () {
      return e.asm.dynCall_viijj.apply(null, arguments);
    };

    e.dynCall_viijjii = function () {
      return e.asm.dynCall_viijjii.apply(null, arguments);
    };

    e.dynCall_viiij = function () {
      return e.asm.dynCall_viiij.apply(null, arguments);
    };

    e.dynCall_viijii = function () {
      return e.asm.dynCall_viijii.apply(null, arguments);
    };

    e.dynCall_iiiiiii = function () {
      return e.asm.dynCall_iiiiiii.apply(null, arguments);
    };

    e.dynCall_iiiiii = function () {
      return e.asm.dynCall_iiiiii.apply(null, arguments);
    };

    e.dynCall_i = function () {
      return e.asm.dynCall_i.apply(null, arguments);
    };

    e.dynCall_iiii = function () {
      return e.asm.dynCall_iiii.apply(null, arguments);
    };

    e.dynCall_iiiii = function () {
      return e.asm.dynCall_iiiii.apply(null, arguments);
    };

    e.dynCall_iiiiiiiiii = function () {
      return e.asm.dynCall_iiiiiiiiii.apply(null, arguments);
    };

    e.dynCall_iiiiiiiii = function () {
      return e.asm.dynCall_iiiiiiiii.apply(null, arguments);
    };

    e.dynCall_viiiiiiiii = function () {
      return e.asm.dynCall_viiiiiiiii.apply(null, arguments);
    };

    e.dynCall_jiji = function () {
      return e.asm.dynCall_jiji.apply(null, arguments);
    };

    e.dynCall_iidiiii = function () {
      return e.asm.dynCall_iidiiii.apply(null, arguments);
    };

    e.dynCall_viiiii = function () {
      return e.asm.dynCall_viiiii.apply(null, arguments);
    };

    e.dynCall_iiiiiiii = function () {
      return e.asm.dynCall_iiiiiiii.apply(null, arguments);
    };

    e.dynCall_iiiiiijj = function () {
      return e.asm.dynCall_iiiiiijj.apply(null, arguments);
    };

    e.dynCall_iiiiij = function () {
      return e.asm.dynCall_iiiiij.apply(null, arguments);
    };

    e.dynCall_iiiiid = function () {
      return e.asm.dynCall_iiiiid.apply(null, arguments);
    };

    e.dynCall_iiiiijj = function () {
      return e.asm.dynCall_iiiiijj.apply(null, arguments);
    };

    e.asm = Xc;
    var Yc;

    e.then = function (a) {
      if (Yc) a(e);else {
        var b = e.onRuntimeInitialized;

        e.onRuntimeInitialized = function () {
          b && b();
          a(e);
        };
      }
      return e;
    };

    function na(a) {
      this.name = "ExitStatus";
      this.message = "Program terminated with exit(" + a + ")";
      this.status = a;
    }

    Sa = function Zc() {
      Yc || ad();
      Yc || (Sa = Zc);
    };

    function ad(a) {
      function b() {
        if (!Yc && (Yc = !0, !va)) {
          Pa = !0;
          Ja(La);
          Ja(Ma);
          if (e.onRuntimeInitialized) e.onRuntimeInitialized();

          if (bd) {
            var b = a;
            b = b || [];
            var d = b.length + 1,
                f = Ba(4 * (d + 1));
            x[f >> 2] = Aa(ca);

            for (var g = 1; g < d; g++) x[(f >> 2) + g] = Aa(b[g - 1]);

            x[(f >> 2) + d] = 0;

            try {
              var h = e._main(d, f);

              if (!ta || 0 !== h) {
                if (!ta && (va = !0, e.onExit)) e.onExit(h);
                da(h, new na(h));
              }
            } catch (l) {
              l instanceof na || ("SimulateInfiniteLoop" == l ? ta = !0 : ((b = l) && "object" === typeof l && l.stack && (b = [l, l.stack]), v("exception thrown: " + b), da(1, l)));
            } finally {}
          }

          if (e.postRun) for ("function" == typeof e.postRun && (e.postRun = [e.postRun]); e.postRun.length;) b = e.postRun.shift(), Oa.unshift(b);
          Ja(Oa);
        }
      }

      a = a || ba;

      if (!(0 < D)) {
        if (e.preRun) for ("function" == typeof e.preRun && (e.preRun = [e.preRun]); e.preRun.length;) Qa();
        Ja(Ka);
        0 < D || (e.setStatus ? (e.setStatus("Running..."), setTimeout(function () {
          setTimeout(function () {
            e.setStatus("");
          }, 1);
          b();
        }, 1)) : b());
      }
    }

    e.run = ad;
    if (e.preInit) for ("function" == typeof e.preInit && (e.preInit = [e.preInit]); 0 < e.preInit.length;) e.preInit.pop()();
    var bd = !0;
    e.noInitialRun && (bd = !1);
    ta = !0;
    ad();
    return load_perspective;
  };
}();

export default load_perspective;
