// @vscode-logging/wrapper/dist/src/settings-changes-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listenToLogSettingsChanges = exports.logLoggerDetails = void 0;
const settings_1 = require("./settings");
function logLoggerDetails(opts) {
    opts.logger.info(`Start Logging in Log Level: <${opts.logLevel}>`);
    opts.logger.info(`Full Logs can be found in the <${opts.logPath}> folder.`);
}
exports.logLoggerDetails = logLoggerDetails;
function listenToLogSettingsChanges(opts) {
    opts.subscriptions.push(opts.onDidChangeConfiguration(e => {
        if (e.affectsConfiguration(opts.loggingLevelProp)) {
            const logLevel = settings_1.getLoggingLevelSetting({
                loggingLevelProp: opts.loggingLevelProp,
                getConfiguration: opts.getConfiguration
            });
            opts.logger.changeLevel(logLevel);
            logLoggerDetails({
                logger: opts.logger,
                logPath: opts.logPath,
                logLevel
            });
        }
    }));
    opts.subscriptions.push(opts.onDidChangeConfiguration(e => {
        if (e.affectsConfiguration(opts.sourceLocationProp)) {
            const newSourceLocationTracking = settings_1.getSourceLocationTrackingSetting({
                sourceLocationProp: opts.sourceLocationProp,
                getConfiguration: opts.getConfiguration
            });
            opts.logger.changeSourceLocationTracking(newSourceLocationTracking);
        }
    }));
}
exports.listenToLogSettingsChanges = listenToLogSettingsChanges;
